MudThemeProvider: Add pseudo css selector for generated theme (5781)

stoepa opened the pull request:
## Description
resolves #5781
resolves #1394

## How Has This Been Tested?
<!-- All PR's should implement unit tests if possible -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

theunis-allegiance committed:
MudThemeProvider: Added sudo css selector for generated theme (5781)

stoepa commented:
Will resolve #5781

stoepa commented:
Also added the functionality to the documentation site. Not sure if this is allowed.

JonBunator commented on src/MudBlazor.Docs/Pages/Customization/Theming/SudoCss.razor:
```diff
@@ -0,0 +1,22 @@
+        </DocsPageSection>
+        <DocsPageSection>
+            <MudDivider Class="my-16" Style="opacity:0;" />
+            @*You are cool like a swimming pool!*@
```
haha

stoepa replied:
Not sure who put that there.

JonBunator reviewed (approved):
Thanks for the PR. Code looks good to me. Imo, nothing speaks against this feature.
One question: Wouldn't it make more sense to add a string prop to allow to set a custom sudo class. (not sure if there are usecases for this)

stoepa commented:
> Thanks for the PR. Code looks good to me. Imo, nothing speaks against this feature. One question: Wouldn't it make more sense to add a string prop to allow to set a custom sudo class. (not sure if there are usecases for this)

It does definitely make sense, but I kept to Enums, to ensure that no invalid sudo classes can be used. Might be able to do some work in terms of allowing whatever is in the Enum or providing your own value but will then update the method to specifically warn the user against it, as it might cause breaking changes. Like a warning, do this at own risk? Will create another pull request hopefully soon for this.

stoepa committed:
Merge pull request #1 from MudBlazor/dev

PR of new changes

theunis-allegiance committed:
Merge branch 'dev' into feature/sudo-css-theme-provider

theunis-allegiance committed:
MudTheme: Added pseudo css selector for generated theme (5781)

stoepa commented:
I have altered how this works a bit. I included in the MudTheme instead of taking in a parameter in the MudThemeProvider.

I have also removed the enum so that you can pass in whatever you want, default will always be :root

I have updated and added additional tests.

I have also renamed from Sudo CSS to Pseudo as it is correctly referred to in the Mozilla developer documentation.

Let me know if you need me to change anything additionaly.

just-the-benno reviewed (commented):
Thanks @stoepa for this feature. I also don't have any objections

just-the-benno commented:
@henon What do you think?

henon commented:
From the documentation I honestly don't know what can be done with this (I can only guess). So for heavily CSS-ignorant users (like me) it would be awesome to have a little example that demonstrates a use-case.

MarianSWA commented:
I think this is very useful, as I want to have for example 2 sections in my page with 2 different themes, so I would change the root selector with a new selector that contains both sections, or root for the main theme and another section with a specific selector for the "inner" section theme. I currently achieve this by deriving from the theme provider and replacing the :root string with my own selector, but it's hacky and I could definitely use this feature to replace it. 

There would also be an usecase (improvement for this feature) that would somehow link the popovers, tooltips and modals to that specific selector. Since they are rendered outside of the div of their trigger component, there is no way to know which of the popovers or menus that are just rendered all together on the bottom of the page belong to the section with the first theme, or the other section with the other theme. For this we would have to know from which section that menu/popover is opened. To have a truly multi theme on the same page support this would be required, but until then modifying the selector is definitely a useful feature.

henon commented:
Thanks @stoepa, sorry for the long wait, I first waited for a response and when it didn't come I forgot about this.

theunis-allegiance commented:
Thanks @henon. No problem about the wait.

POnakS commented:
@stoepa @henon 
I think it would make sense to make default selector `:root, :host`. 
The issue is that `:root` used by MudTheme is basically html element. Shadow dom however, does not have html element and it has special selector `:host` to handle this use case (don't ask me why, I'm not the one who designed this mess :)). FYI Shadow dom is used to separate CSS (if you embed blazor app in legacy app).

After change introduced by this MR if you embed Mud component inside shadow dom you have to educate yourself about how this works inside shadow dom, and I think Mud could handle this out of the box to save time for the users (especially considering fact that competition like Radzen does not have such issues, so someone may switch framework just because it does not work OOTB).

POnakS commented:
@stoepa I see you're working with web components, do you have solution for events (i.e. onchange) not working inside shadow dom?
